### PR TITLE
Add SoftwareReport Archive json generator

### PR DESCRIPTION
**Pull request by bogdan-damian-bgd**
# Description
## Generation of additional SoftwareReport output .json archive
Changes provided for **Windows, Linux and macOS images**.

The idea is to keep track of all installed items per image - the essence of the output archive json is that each of the installed items in the image is described with a unique ID and that other properties can be (easily) added for later use as needed.

The initial reason for introducing these changes is because this kind of data source is necessary for further search and data processing, i.e. stable generation of [Release pages](https://github.com/actions/runner-images/releases) in which the items will be sorted into required groups, such as _Updated_, _Added_ and _Deleted_ (see examples [1](https://www.figma.com/file/skugzW7gXhENs6ZKISNaaz/Image-Release?node-id=0%3A1), [2](https://github.com/bogdan-damian-bgd/runner-images/pull/8)).

**
For easy archiving, each installed application is additionally described into its own `ArchiveItem` object/class, which will carry all the properties needed to describe the installed item. When creating such an ArchiveItem, it is now necessary to assign a unique identifier of the installed item.

Follow the links to the example files [Windows2022-Archive-Table.json](https://github.com/bogdan-damian-bgd/runner-images/blob/9124887ebfb1dabbee09878cbf8c35fd02670a5f/images/win/Windows2022-Archive-Table.json) and [Windows2022-Archive_Grouped.json](https://github.com/bogdan-damian-bgd/runner-images/blob/9124887ebfb1dabbee09878cbf8c35fd02670a5f/images/win/Windows2022-Archive_Grouped.json) to find demo samples for two suggested structures of the output json file.

## Check list
- [x] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [x] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1 [hunk 1]
```diff
@@ -175,4 +187,8 @@ function Build-AndroidEnvironmentTable {
             "Value" = if ($shouldResolveLink.Contains($_.Name )) { Get-PathWithLink($_.Value) } else {$_.Value}
         }
     }
+
+    $output | ForEach-Object { $Archive.Add("$($_.Name)|$($_.Value)", "Env_$($_.Name)") } | Out-Null
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1:191]: $output is not defined

File: images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1 [hunk 2]
```diff
@@ -175,4 +187,8 @@ function Build-AndroidEnvironmentTable {
...
+
+    return $output
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1:193]: this line will never execute (see - 184)

File: images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1 [hunk 3]
```diff
@@ -82,6 +86,10 @@ function Build-AndroidTable {
             "Version" = $_.Version
         }
     }
+
+    $output | ForEach-Object { $Archive.Add("$($_."Package Name")|$($_.Version)", "Android_$($_."Package Name")".Replace(" ", "")) } | Out-Null
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.Android.psm1:90]: nitpick:  create a new function for this statement

File: images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1 [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    ArchiveItems() {
+        $this.items = New-Object System.Collections.ArrayList
+        $this.activeHeaders = New-Object string[] 5
+    }
+
+    [string] Add($Title, $Id) {
+        $item = [ArchiveItem]::New()
+        $item.Id = $Id
+        $item.Title = $Title
+        $item.Headers = $this.activeHeaders | ForEach-Object { $_ } | Where-Object { $_.Length -ge 1 }
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1:20]: why do we need ForEach-Object { $_ }?

File: images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1 [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    [string[]] $Headers
+}
+
+class ArchiveItems {
+    hidden [System.Collections.ArrayList] $items
+    hidden [string[]] $activeHeaders
+
+    ArchiveItems() {
+        $this.items = New-Object System.Collections.ArrayList
+        $this.activeHeaders = New-Object string[] 5
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1:13]: What is 5?

File: images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1 [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        $this.activeHeaders[$Level-1] = $Name
+        for ($i = $Level; $i -lt $this.activeHeaders.Length; $i++) {
+            $this.activeHeaders[$i] = ""
+        }
+        return $Name
+    }
+
+    [string] ToJson() {
+        $final = @()
+        foreach ($item in $this.items) {
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1:45]: nitpick: use $final = $this.items | Foreach ...

File: images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1 [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                Id = $item.Id
+                Title = $item.Title
+                Headers = $headers
+            }
+        }
+        return ConvertTo-Json $final -Depth 10
+    }
+
+    [string] ToJsonGrouped() {
+        [string[]] $lastHeaders = ($this.items[0]).Headers
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1:59]: remove ()

File: images/linux/scripts/SoftwareReport/SoftwareReport.CachedTools.psm1 [hunk 8]
```diff
@@ -1,35 +1,104 @@
 function Get-ToolcacheRubyVersions {
+    param (
+        [ArchiveItems] $Archive
+    )
+
     $toolcachePath = Join-Path $env:AGENT_TOOLSDIRECTORY "Ruby"
-    return Get-ChildItem $toolcachePath -Name | Sort-Object { [Version]$_ }
+    $output = Get-ChildItem $toolcachePath -Name | Sort-Object { [Version]$_ }
+    $output | ForEach-Object {
```
Comment (al-cheb) [inline images/linux/scripts/SoftwareReport/SoftwareReport.CachedTools.psm1:8]: create function to stop duplicate the same code

File: images/win/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1 [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        $this.activeHeaders[$Level-1] = $Name
+        for ($i = $Level; $i -lt $this.activeHeaders.Length; $i++) {
+            $this.activeHeaders[$i] = ""
+        }
+        return $Name
+    }
+
+    [string] ToJson() {
```
Comment (al-cheb) [inline images/win/scripts/SoftwareReport/SoftwareReport.ArchiveItem.ps1:43]: Where do we use this method?

**Review by slumericanB — APPROVED** [review]
Good

**Review by slumericanB — APPROVED** [review]
Good

**Pull request closed without merging**